### PR TITLE
Local costmap in "odom" frame instead of  "map" frame (SS-415 SS-456 BTB-181)

### PR DESCRIPTION
**Pull request by belem2050**
## Description
Local costmap should be in "odom" frame not "map" frame
## Purpose
The purpose here is to change it from map to odom
## Dependency
[Local costmap param](https://github.com/cmrobotics/serena_bringup/pull/171)

### CODE REVIEW

File: nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp [hunk 1]
```diff
@@ -508,8 +508,10 @@ bool RegulatedPurePursuitController::isCollisionImminentExtendedSearch()
     std::back_inserter(extended_collision_check_path.poses),
     [](geometry_msgs::msg::PoseStamped & poseStamped) {
       // Modify Z For visualization in rviz so it is drawn between global plan and collision arc
+      geometry_msgs::msg::PoseStamped transformed_pose;
       poseStamped.pose.position.z = 0.005;
-      return poseStamped;
+      transformPose(costmap_ros_->getBaseFrameID(), poseStamped, transformed_pose);
```
Comment (dimaxano) [inline nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp]: Why do we transform it to the base frame, not to the global (which is `odom` after bringup is merged)

Comment (TanmayDeshmukh) [reply nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp]: Good point.. I wonder if collision checking that happens in the later lines works either way

Comment (belem2050) [reply nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp]: > Why do we transform it to the base frame, not to the global (which is `odom` after bringup is merged)

Actually, I did not get the good understanding of the function ``` transformPose ```.

Comment (belem2050) [reply nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp]: > Good point.. I wonder if collision checking that happens in the later lines works either way

I checked and was looking not bad but I´d like to know more.

**Comment by TanmayDeshmukh** [comment]
I will convert this to draft in preparation for freezing development

File: nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp [hunk 2]
```diff
@@ -506,13 +506,17 @@ bool RegulatedPurePursuitController::isCollisionImminentExtendedSearch()
     global_plan_.poses.begin(),
     last_pose_it,
     std::back_inserter(extended_collision_check_path.poses),
-    [](geometry_msgs::msg::PoseStamped & poseStamped) {
+    [&](geometry_msgs::msg::PoseStamped & poseStamped) {
```
Comment (TanmayDeshmukh) [inline nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp]: Try passing only the objects you need 
`[global_plan_, costmap_ros_](geometry_msgs::msg::PoseStamped & poseStamped) `

or something like that

File: nav2_regulated_pure_pursuit_controller/test/test_regulated_pp.cpp [hunk 3]
```diff
@@ -642,6 +642,105 @@ TEST(RegulatedPurePursuitTest, testDynamicParameter)
...
+
+
+  unsigned int width = costmap->getLayeredCostmap()->getCostmap()->getSizeInCellsX();
+  unsigned int height = costmap->getLayeredCostmap()->getCostmap()->getSizeInCellsY();
+
+  // collision should be imminent for inscribed_inflated and lethal but not for free space
+
+  for (unsigned int x = 0; x < width; ++x) {
+      for (unsigned int y = 0; y < height; ++y) {
+          costmap->getLayeredCostmap()->getCostmap()->setCost(x, y, nav2_costmap_2d::LETHAL_OBSTACLE);
```
Comment (TanmayDeshmukh) [inline nav2_regulated_pure_pursuit_controller/test/test_regulated_pp.cpp]: Could be a more meaningful/ strict-er unit test if you could test that a part of the path is in collision and part of it is not. 
So for example, if someone were to change if_any() to all_of() inside `isCollisionImminentExtendedSearch()` these tests would succeed, but we want them to fail.

File: nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp [hunk 4]
```diff
@@ -506,13 +506,17 @@ bool RegulatedPurePursuitController::isCollisionImminentExtendedSearch()
...
-      return poseStamped;
+      geometry_msgs::msg::PoseStamped transformed_pose;
+      poseStamped.header = global_plan_.header;
+      transformPose(costmap_ros_->getGlobalFrameID(), poseStamped, transformed_pose);
+      transformed_pose.pose.position.z = 0.005;
+      return transformed_pose;
     }
   );
-  extended_collision_check_path.header = global_plan_.header;
+  extended_collision_check_path.header.frame_id = costmap_ros_->getGlobalFrameID();
```
Comment (fvlvivs) [inline nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp:518]: Do not repeat: assign `costmap_ros_->getGlobalFrameID()` to a variable.

Comment (belem2050) [reply nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp:518]: I got you. Basically I could. ``` costmap_ros``` already is a variable and I do not know if it is better to create a new one based on it.

**Review by TanmayDeshmukh — APPROVED** [review]

**Comment by dimaxano** [comment]
@BELEMMoumouni 
Was it tested on the robot?

**Comment by belem2050** [comment]
> @BELEMMoumouni Was it tested on the robot?

Singlely, tested, no cllision but A deeper test by using multiple robots and fleet will be done when what I am working on now is finished.

**Review by dimaxano — APPROVED** [review]

**Comment by TanmayDeshmukh** [comment]
Oh actually we should've waited until we merge changes for[ this issue](https://cm-robotics.atlassian.net/jira/software/projects/SS/boards/48?selectedIssue=SS-507) because this breaks that..

Oh well

**Comment by belem2050** [comment]
> Oh actually we should've waited until we merge changes for[ this issue](https://cm-robotics.atlassian.net/jira/software/projects/SS/boards/48?selectedIssue=SS-507) because this breaks that..
> 
> Oh well
Oh yes, you're right. I just thought ```pre-release``` would have been fine.

**Pull request merged**